### ISSUE TITLE
Inform user of local community events

### ISSUE THREAD

**Issue by westnordost** [westnordost/StreetComplete#1798]
The app knows the user's location and by subscribing to the feed from https://github.com/thomersch/openstreetmap-calendar (or the calendar in the wiki), using the new notification system, the app could notify the user of local community events.

Optionally, the user should be able to completely unsubscribe from this in the settings.
Optionally, the user should be able to add the event to the calendar on his smartphone with one click.

The notification button should only be used to hint to new events, what is probably necessary is to implement some kind of calendar view which would then pop up and show all the events. This might be too much work for what its worth though, as usually, in one city, there is usually also only just one recurring event (i.e. Stammtisch).

Before this is started, it must be ascertained if openstreetmap-calendar actually provides the lat-lon of the event, or just the city name. If the latter, a geocoder mus be involved, making the whole thing more complicated to implement.

**Comment by westnordost** [comment]
I submitted an [application for a microgrant at the OSMF](https://wiki.openstreetmap.org/wiki/Microgrants/Microgrants_2020/Proposal/StreetComplete_as_an_entry_point_to_OpenStreetMap) to develop this feature.

**Comment by kmpoppe** [comment]
@westnordost Did you, above writing it in the microgrant application and that you asserted above, have any other mockups for an implementation?

**Comment by westnordost** [comment]
You probably mean mockups in terms of UI? No. Are you interested in taking on that task?

### Some thoughts:

#### Message
- should only display nearby community events. 25km or so radius should be plenty. (New users likely don't bring the motivation to travel that far for a not-even-yet-really-newfound-hobby)
- nearby community events should appear as a message in the message area
- ideally, they should either rather pop up after uploading changes rather than on download of map data or not pop up right away for new users. I.e. they should pop up after the user already engaged with the app, logged in etc.
- the message should pop up looking nice and with a nice animation. E.g. as a tear-off calendar, prominently displaying the date, then the name+location of the event with one button that would lead to details of the event and/or an intent to add this to your device's calendar. The appear-animation could feature some rapidly flipping through the calendar to arrive at that page or so or otherwise an animation that looks like the calendar is opened.
- new events should of course only be notified about once, so it must be persisted somehow until which event we already read the events feed.
- likewise, if users are not interested in these events, it should be possible to turn these off in the settings. Maybe also for recurring events, like e.g. if the user is just not interested to go to a particular recurring Stammtisch. All in all, it should not start to get annoying for the user

#### Tab in user screen
- There could also be a tab in the user screen where the user could browse all (local?) events. Maybe he dismissed this one event but would like to look it up later. (By the way, material design suggests to use sidebar navigation over tabbed navigation with subpages > 4, IIRC). Alternatively, if the default osmcal UI is nice and fit for mobile use, one could also just link the calendar website (with some parameters?) or display it in a webview.
- It should be possible from within the app to create own new events.

**Comment by kmpoppe** [comment]
While I was browsing through old Issues I found this one, and - as I already found this interesting when you made the microgrant application - I thought of investigating further.

So yes, if you like, you can assign this one to me.

Thoughts on your thoughts:

* I also put my hands on the osmcal source, trying to incorporate a radius instead of the fixed 50km, not quite there yet, though.

* There are no recurring events per definition, so ignoring a special event would depend on the creator to use the exact same phrasing every time.

* I'm not sure I would ever be able to implement your creative UI vision (exhibit A: Boat Rental Icons), but I'm happy to figure out how the backend might work.

**Comment by westnordost** [comment]
Right. At least the UX somewhat dictates how the backend should look like. So, a simple dialog that just contains all the functionality would be fine at first, and then maybe I jump in to make the interface more colorful. In case you start working on it soon, please base your (UI) work on the `compose-edithistory` branch.

**Comment by matkoniecz** [comment]
> So yes, if you like, you can assign this one to me.

Based on my own "lets start 136 projects at once" it seems better to first finish first one of #5889 #5903 #5741

**Comment by kmpoppe** [comment]
Whereof one is finished and awaits a beta 59.1 or 60, one needs tweaks that I can't finish _right now_ because I'm on holiday and in one I explicitly asked for help with icons as I am unable to create something @westnordost envisioned.

And yes, I'm a busy adult that has other things to do that pay my bills. As almost everyone in this project.

**Comment by matkoniecz** [comment]
In cuch case I am sorry for doubting and overly matching to my situation.

**Comment by kmpoppe** [comment]
More food for thought:

* Dynamic filter is now in osmcal. We can therefore now display events in a 25k radius.
* Events don't have a creation date in the API output and @thomersch thinks, that we do not need it because
  * There is a `days` filter, so we can only display X amount of days in the future (I think 90 would be a good value)
  * The id of the event is counting upwards. As the filter might be displaying events at different places (the user moves around the world using the app), I think the best way to know what we have displayed is storing the ids in a local db (I will need to dive into the code finding out how that works ;-))

**Comment by westnordost** [comment]
For single persisted data, check out

https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/data/preferences/Preferences.kt

**Comment by tordans** [comment]
> 1. Notify users of upcoming events from [OSM calendar](https://osmcal.org/) in their vicinity ([#1798](https://github.com/streetcomplete/StreetComplete/issues/1798))
> ([Source](https://github.com/osmfoundation/ewg_bidding/pull/50/files))

I wonder if we need to improve osmcal to make this work properly.
We have quite a few events that are region specific, like "Berlin", "Brandenburg" (State), "Germany" (Country).
But osmcal only allows to set a point. 
For our german-wide remote events we just put the pin somewhere but that is always wrong.

This is also an issue with osmcal for the RSS Feed, because I cannot subscribe to "Germany", just "all" and "near me" where the latest will not catch all German events and the first (all) will show irrelevant stuff.

There are issues on those topics in the osmcal repo like the idea of groups / tags / categories. But maybe we need something like an event-area maybe based on country coder or a custom geometry?

**Comment by westnordost** [comment]
I don't think that OsmCal need to be changed to query for local events properly. StreetComplete can just query **all** events in regular intervals (e.g. once per start of the app or similar) and then filter by distance itself. The resulting JSON is not that big. I just checked, it is about 7kB over the wire.

I think a default max distance of 20km to events from the current location would be a reasonable default.

If it is possible to filter out remote events, that would be nice though. (IMO they are not relevant for this use case.) It seems there is no field for this.

**Comment by mnalis** [comment]
> This is also an issue with osmcal for the RSS Feed, because I cannot subscribe to "Germany", 

Uh, can you not?  What does https://osmcal.org/events.rss?in=Germany do then?

(for me https://osmcal.org/subscribe/?in=Croatia always seemed to work just fine; i.e. only gave events in Croatia, and never those outside Croatia. But maybe I was just lucky? Though I use iCalendar not RSS subscription, but I assume they should filter the same way)
